concordium-consensus/src/Concordium/GlobalState/Persistent/LMDB.hs

```diff
@@ -688,3 +693,44 @@ writeFinalizationComposite finRec blocktss = resizeOnFull (finRecSize + blocksSi
           digestSize + 64 + digestSize + (32 * Prelude.length vs) + 48 + 64
     blocksSize = sum . map (\b -> 2*digestSize + fromIntegral (LBS.length . fst . fst $ b)) $ serializedBlocksTss
     tssSize = (* (2 * digestSize + 16)) . sum . map (Prelude.length . snd) $ blocktss
+
+-- If an explicitly finalized block does not satisfy a predicate, delete it, together with its
+-- finalization record, all blocks implicitly finalized by its finalization records and all
+-- associated transaction statuses. Returns the last finalized block hash available in the unrolled
+-- treestate/blockstate on success.
+unrollTreeStateWhile :: (MonadLogger m, MonadIO m, IsProtocolVersion pv, FixedSizeSerialization st) =>
+  DatabaseHandlers pv st -> (StoredBlock pv st -> IO Bool) -> m (Either String (FinalizationRecord, StoredBlock pv st))
+unrollTreeStateWhile dbh p = resizeOnFullInternal 4096 dbh $ \h -> transaction (h ^. storeEnv) False $ \txn ->
+  let loopBlocks finIndex sb | NormalBlock block <- sbBlock sb =
+        do
+          let bh = _bpHash . sbInfo $ sb
+          isP <- p sb
+          if isP then do
+            delBlockOK <- deleteRecord txn (h ^. blockStore) bh
+            delBlockHashOK <- deleteRecord txn (h ^. finalizedByHeightStore) (_bpHeight . sbInfo $ sb)
+            delTxsOK <- mapM (deleteRecord txn (h ^. transactionStatusStore)) ((wmdHash <$>) . blockTransactions $ block)
+            if delBlockOK && delBlockHashOK && and delTxsOK then do
+              -- Before deleting a finalization record for a block, we must ensure that its ancestors
+              -- are not finalized by the same finalization record. We only delete the finalization
+              -- record after finding an ancestor block with a different finalization index. In
+              -- particular, we delete the finalization record if the parent block is the genesis
+              -- block which, by definition, is the only block with the finalization index 0.
+              loadRecord txn (h ^. blockStore) (blockPointer block) >>= \case
+                Just sbParent | finIndex == sbFinalizationIndex sbParent -> loopBlocks finIndex sbParent
+                Just sbParent -> do
+                  delFinRecOK <- deleteRecord txn (h ^. finalizationRecordStore) finIndex
+                  if delFinRecOK then loopBlocks (sbFinalizationIndex sbParent) sbParent
+                    else return . Left $ "Could not delete the finalization record with index "
+                    <> show finIndex <> " finalizing block " <> show bh
+                Nothing -> return . Left $ "Could not read the parent of the block " <> show bh
+             else return . Left $ "Could not rollback block " <> show bh
+          else returnBlock finIndex sb
+      -- genesis blocks won't be deleted
```

Comment:
Perhaps rationale as to why?

To be clear,I think it is acceptable to not handle this specific case where the genesis state is corrupted since it is a bit more tricky to handle, but just insert a comment as to why we chose it.